File: projects/api-testing/tests/playwright-csharp/AbmCrudAcTests.cs
using System.Net;
using System.Text.Json;
using Microsoft.Playwright;
using NUnit.Framework;

namespace QaPortfolio.PlaywrightApiTests;

[TestFixture]
public class AbmCrudAcTests
{
    private string? _createdId;

    [Test, Order(1)]
    public async Task AC001_CreateValidItem()
    {
        var response = await ApiTestFixture.Request.PostAsync("/api/items", new APIRequestContextOptions
        {
            DataObject = new { name = "Item QA-001", description = "Playwright", active = true }
        });
        Assert.That(response.Status, Is.EqualTo(201));
        var json = await response.JsonAsync();
        _createdId = json?.GetProperty("id").GetString();
        Assert.That(_createdId, Is.Not.Null.And.Not.Empty);
    }

    [Test, Order(2)]
    public async Task AC002_CreateInvalidName()
    {
        var response = await ApiTestFixture.Request.PostAsync("/api/items", new APIRequestContextOptions
        {
            DataObject = new { name = "" }
        });
        Assert.That(response.Status, Is.EqualTo(400));
    }

    [Test, Order(3)]
    public async Task AC003_ListPaged()
    {
        var response = await ApiTestFixture.Request.GetAsync("/api/items?page=0&size=10");
        Assert.That(response.Status, Is.EqualTo(200));
        var body = await response.JsonAsync();
        Assert.That(body!.TryGetProperty("content", out _), Is.True);
    }

    [Test, Order(4)]
    public async Task AC004_FilterActive()
    {
        var response = await ApiTestFixture.Request.GetAsync("/api/items?active=true");
        Assert.That(response.Status, Is.EqualTo(200));
    }

    [Test, Order(5)]
    public async Task AC005_GetById()
    {
        Assert.That(_createdId, Is.Not.Null);
        var response = await ApiTestFixture.Request.GetAsync($"/api/items/{_createdId}");
        Assert.That(response.Status, Is.EqualTo(200));
    }

    [Test, Order(6)]
    public async Task AC006_GetNotFound()
    {
        var response = await ApiTestFixture.Request.GetAsync(
            "/api/items/00000000-0000-0000-0000-000000000099");
        Assert.That(response.Status, Is.EqualTo(404));
    }

    [Test, Order(7)]
    public async Task AC007_UpdateItem()
    {
        var id = await CreateItemAsync();
        var response = await ApiTestFixture.Request.PutAsync($"/api/items/{id}", new APIRequestContextOptions
        {
            DataObject = new { name = "Item QA-007-up", active = false }
        });
        Assert.That(response.Status, Is.EqualTo(200));
        var json = await response.JsonAsync();
        Assert.That(json!.GetProperty("name").GetString(), Is.EqualTo("Item QA-007-up"));
    }

    [Test, Order(8)]
    public async Task AC008_UpdateNotFound()
    {
        var response = await ApiTestFixture.Request.PutAsync(
            "/api/items/00000000-0000-0000-0000-000000000099",
            new APIRequestContextOptions { DataObject = new { name = "X" } });
        Assert.That(response.Status, Is.EqualTo(404));
    }

    [Test, Order(9)]
    public async Task AC009_DeleteItem()
    {
        var id = await CreateItemAsync();
        var response = await ApiTestFixture.Request.DeleteAsync($"/api/items/{id}");
        Assert.That(response.Status, Is.EqualTo(204));
    }

    [Test, Order(10)]
    public async Task AC010_DeleteNotFound()
    {
        var response = await ApiTestFixture.Request.DeleteAsync(
            "/api/items/00000000-0000-0000-0000-000000000099");
        Assert.That(response.Status, Is.EqualTo(404));
    }

    private static async Task<string> CreateItemAsync()
    {
        var response = await ApiTestFixture.Request.PostAsync("/api/items", new APIRequestContextOptions
        {
            DataObject = new { name = $"Temp-{Guid.NewGuid()}", active = true }
        });
        Assert.That(response.Status, Is.EqualTo(201));
        var json = await response.JsonAsync();
        return json!.GetProperty("id").GetString()!;
    }
}

File: projects/api-testing/tests/rest-assured-java/src/test/java/com/qaportfolio/api/AbmCrudAcTests.java
package com.qaportfolio.api;

import io.restassured.http.ContentType;
import org.junit.jupiter.api.*;

import java.util.Map;
import java.util.UUID;

import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.*;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class AbmCrudAcTests extends BaseApiTest {
    private static String createdId;

    @Test
    @Order(1)
    @DisplayName("AC-001 Create valid item")
    void ac001_createItem() {
        createdId = given()
                .contentType(ContentType.JSON)
                .body(Map.of(
                        "name", "Item QA-001",
                        "description", "Rest-Assured",
                        "active", true))
                .when()
                .post("/api/items")
                .then()
                .statusCode(201)
                .body("id", notNullValue())
                .body("name", equalTo("Item QA-001"))
                .extract()
                .path("id");
    }

    @Test
    @Order(2)
    @DisplayName("AC-002 Create invalid name")
    void ac002_createInvalid() {
        given()
                .contentType(ContentType.JSON)
                .body(Map.of("name", ""))
                .when()
                .post("/api/items")
                .then()
                .statusCode(400)
                .body("message", notNullValue());
    }

    @Test
    @Order(3)
    @DisplayName("AC-003 List paged")
    void ac003_listPaged() {
        given()
                .queryParam("page", 0)
                .queryParam("size", 10)
                .when()
                .get("/api/items")
                .then()
                .statusCode(200)
                .body("content", notNullValue())
                .body("totalElements", greaterThanOrEqualTo(0));
    }

    @Test
    @Order(4)
    @DisplayName("AC-004 Filter active")
    void ac004_filterActive() {
        createFreshItem();
        given()
                .queryParam("active", true)
                .when()
                .get("/api/items")
                .then()
                .statusCode(200)
                .body("content", notNullValue());
    }

    @Test
    @Order(5)
    @DisplayName("AC-005 Get by id")
    void ac005_getById() {
        Assumptions.assumeTrue(createdId != null);
        given()
                .when()
                .get("/api/items/" + createdId)
                .then()
                .statusCode(200)
                .body("id", equalTo(createdId));
    }

    @Test
    @Order(6)
    @DisplayName("AC-006 Get not found")
    void ac006_getNotFound() {
        given()
                .when()
                .get("/api/items/00000000-0000-0000-0000-000000000099")
                .then()
                .statusCode(404);
    }

    @Test
    @Order(7)
    @DisplayName("AC-007 Update item")
    void ac007_update() {
        String id = createFreshItem();
        given()
                .contentType(ContentType.JSON)
                .body(Map.of("name", "Item QA-007-up", "active", false))
                .when()
                .put("/api/items/" + id)
                .then()
                .statusCode(200)
                .body("name", equalTo("Item QA-007-up"));
    }

    @Test
    @Order(8)
    @DisplayName("AC-008 Update not found")
    void ac008_updateNotFound() {
        given()
                .contentType(ContentType.JSON)
                .body(Map.of("name", "X"))
                .when()
                .put("/api/items/00000000-0000-0000-0000-000000000099")
                .then()
                .statusCode(404);
    }

    @Test
    @Order(9)
    @DisplayName("AC-009 Delete item")
    void ac009_delete() {
        String id = createFreshItem();
        given().when().delete("/api/items/" + id).then().statusCode(204);
    }

    @Test
    @Order(10)
    @DisplayName("AC-010 Delete not found")
    void ac010_deleteNotFound() {
        given()
                .when()
                .delete("/api/items/00000000-0000-0000-0000-000000000099")
                .then()
                .statusCode(404);
    }

    private static String createFreshItem() {
        return given()
                .contentType(ContentType.JSON)
                .body(Map.of("name", "Temp-" + UUID.randomUUID(), "active", true))
                .when()
                .post("/api/items")
                .then()
                .statusCode(201)
                .extract()
                .path("id");
    }
}
